File: gz_dsp.c
#include "Python.h"
#include "arrayobject.h"
#include "gz_dsp.c"
#include <math.h>

static PyMethodDef gz_dsp_methods {
	{"gz_dsp", gz_dsp, METH_VARARGS, "Calculate the intensity of a particular frequency. Args are (samples, freq) \
	samples is a 1d numpy array."}
	{NULL, NULL, 0, NULL}
};

float goertzel(float freq, int size, float x[])
{
	int i;
	float coeff;
	float s, s_prev1 = 0.0f, s_prev2 = 0.0f;

	coeff = 2.0f * cosf(2.0f * M_PI * freq);

	for (i = 0; i < size; i++) {
		s = x[i] + (coeff * s_prev1) - s_prev2;
		s_prev2 = s_prev1;
		s_prev1 = s;
	}

	return (s_prev1 * s_prev1) + (s_prev2 * s_prev2)
		 - (s_prev1 * s_prev2 * coeff);
}

File: gz_dsp.py
import numpy
import math
import morse_discriminate

# implementation of goertzel function
def gz_dsp(matrix, freq):
	normalized_freq = freq/morse_discriminate.SAMPLE_FREQ
	coeff = 2 * math.cos(2 * math.pi * normalized_freq)
	s = [0, 0, 0]
	for sample in matrix:
		sample = float(sample)
		s[0] = sample + (coeff * s[1]) - s[2]
		s[1:3] = s[0:2]
	return (s[1]**2) + (s[2]**2) - (s[1] * s[2] * coeff)
